src/workerd/jsg/fast-api.h

```diff
@@ -0,0 +1,131 @@
+#pragma once
+
+#include <workerd/jsg/web-idl.h>
+
+#include <v8-fast-api-calls.h>
+#include <v8-local-handle.h>
+#include <v8-value.h>
+
+#include <kj/common.h>
+
+namespace workerd::jsg {
+
+template <typename T>
+struct IsFastApiCompatibleWrappedType1 {
+  static constexpr bool value = webidl::isDictionaryType<T> ||
+      webidl::isNonCallbackInterfaceType<T> || kj::isSameType<T, kj::String>() ||
+      kj::isSameType<T, jsg::BufferSource>();
+};
+
+template <typename T>
+struct IsFastApiMaybe {
+  static constexpr bool value = false;
+};
+
+template <typename U>
+struct IsFastApiMaybe<kj::Maybe<U>> {
+  static constexpr bool value = IsFastApiCompatibleWrappedType1<U>::value;
+};
+
+template <typename U>
+struct IsFastApiMaybe<jsg::Optional<U>> {
+  static constexpr bool value = IsFastApiCompatibleWrappedType1<U>::value;
+};
+
+template <typename U>
+struct IsFastApiMaybe<jsg::LenientOptional<U>> {
+  static constexpr bool value = IsFastApiCompatibleWrappedType1<U>::value;
+};
+
+template <typename T>
+struct IsFastApiCompatibleWrappedType {
+  static constexpr bool value =
+      IsFastApiCompatibleWrappedType1<T>::value || IsFastApiMaybe<T>::value;
+};
+
+// Helper to determine if a type can be used as a return type in V8 Fast API
+template <typename T>
+struct IsFastApiCompatibleReturnType {
+  static constexpr bool value = kj::isSameType<T, void>() || kj::isSameType<T, bool>() ||
+      kj::isSameType<T, int32_t>() || kj::isSameType<T, uint32_t>() || kj::isSameType<T, float>() ||
+      kj::isSameType<T, double>();
+};
+
+// Helper to determine if a type can be used as a parameter in V8 Fast API
+template <typename T>
+struct IsFastApiCompatibleParamType {
+  static constexpr bool value = IsFastApiCompatibleReturnType<T>::value ||
+      kj::isSameType<T, v8::Local<v8::Value>>() || kj::isSameType<T, v8::Local<v8::Object>>() ||
+      IsFastApiCompatibleWrappedType<T>::value;
+};
+
+// Specialization for jsg::Lock& - this is compatible with Fast API.
+// This allows methods that take jsg::Lock& as their first parameter to be used with V8 Fast API.
+// The Lock will be extracted from the V8 isolate in the fast path, enabling methods that need
+// access to the V8 context and isolate to still benefit from the performance of Fast API.
+template <>
+struct IsFastApiCompatibleParamType<jsg::Lock&> {
+  static constexpr bool value = true;
+};
```

Comment:
Is it not possible to handle this using kj::isSameType and merge into the earlier specialization?